libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -44,3 +44,54 @@ int32_t AP_BattMonitor_Backend::get_capacity() const
 {
     return _mon.pack_capacity_mah(_state.instance);
 }
+
+// update battery resistance estimate
+// faster rates of change of the current and voltage readings cause faster updates to the resistance estimate
+// the battery resistance is calculated by comparing the latest current and voltage readings to a low-pass filtered current and voltage
+// high current steps are integrated into the resistance estimate by varying the time constant of the resistance filter
+void AP_BattMonitor_Backend::update_resistance_estimate()
+{
+    // return immediately if no current
+    if (!has_current() || _state.current_amps <= 0.0f) {
+        return;
+    }
+
+    // update maximum current seen since startup and protect against divide by zero
+    _current_max_amps = MAX(_current_max_amps, _state.current_amps);
+    float current_delta = _state.current_amps - _current_filt_amps;
+    if (is_zero(_current_max_amps) || is_zero(current_delta)) {
+        return;
+    }
+
+    // update reference voltage and current
+    if (_state.voltage > _resistance_voltage_ref) {
+        _resistance_voltage_ref = _state.voltage;
+        _resistance_current_ref = _state.current_amps;
+    }
+
+    // calculate time since last update
+    uint32_t now = AP_HAL::millis();
+    float loop_interval = (now - _resistance_timer_ms) / 1000.0f;
+    _resistance_timer_ms = now;
+
+    // estimate short-term resistance
+    float filt_alpha = constrain_float(loop_interval/(loop_interval + AP_BATT_MONITOR_RES_EST_TC_1), 0.0f, 0.5f);
+    float resistance_alpha = MIN(1, AP_BATT_MONITOR_RES_EST_TC_2*fabsf((_state.current_amps-_current_filt_amps)/_current_max_amps));
+    float resistance_estimate = -(_state.voltage-_voltage_filt)/current_delta;
+    if (resistance_estimate > 0){
```

Comment:
```if (is_positive(resistance_estimate))``` would be better, you want to avoid tiny floats that would otherwise fail the is_zero check.